vmm/src/igvm/igvm_loader.rs

```diff
@@ -0,0 +1,447 @@
+    #[error("allocate address space")]
+    MemoryManager(MemoryManagerError),
+    #[error("Error modifying host access to isolated pages: {0}")]
+    ModifyHostAccess(#[source] hypervisor::HypervisorVmError),
+    #[error("Error importing isolated pages: {0}")]
+    ImportIsolatedPages(#[source] hypervisor::HypervisorVmError),
+    #[error("Error completing importing isolated pages: {0}")]
+    CompleteIsolatedImport(#[source] hypervisor::HypervisorVmError),
+}
+
+fn memory_map_entry(range: &ArchMemRegion) -> IGVM_VHS_MEMORY_MAP_ENTRY {
+    assert!(range.size as u64 % HV_PAGE_SIZE == 0);
+    if range.r_type == RegionType::Ram {
+        IGVM_VHS_MEMORY_MAP_ENTRY {
+            starting_gpa_page_number: range.base / HV_PAGE_SIZE,
+            number_of_pages: range.size as u64 / HV_PAGE_SIZE,
+            entry_type: MemoryMapEntryType::MEMORY,
+            flags: 0,
+            reserved: 0,
+        }
+    } else {
+        IGVM_VHS_MEMORY_MAP_ENTRY {
+            starting_gpa_page_number: range.base / HV_PAGE_SIZE,
+            number_of_pages: range.size as u64 / HV_PAGE_SIZE,
+            entry_type: MemoryMapEntryType::PLATFORM_RESERVED,
+            flags: 0,
+            reserved: 0,
+        }
+    }
+}
+
+#[allow(dead_code)]
+#[derive(Copy, Clone)]
+struct GpaPages {
+    pub gpa: u64,
+    pub page_type: u32,
+    pub page_size: u32,
+}
+
+#[derive(Debug)]
+enum ParameterAreaState {
+    /// Parameter area has been declared via a ParameterArea header.
+    Allocated { data: Vec<u8>, max_size: u64 },
+    /// Parameter area inserted and invalid to use.
+    Inserted,
+}
+
+// Import a parameter to the given parameter area.
+fn import_parameter(
+    parameter_areas: &mut HashMap<u32, ParameterAreaState>,
+    info: &IGVM_VHS_PARAMETER,
+    parameter: &[u8],
+) -> Result<(), Error> {
+    let (parameter_area, max_size) = match parameter_areas
+        .get_mut(&info.parameter_area_index)
+        .expect("parameter area should be present")
+    {
+        ParameterAreaState::Allocated { data, max_size } => (data, max_size),
+        ParameterAreaState::Inserted => panic!("igvmfile is not valid"),
+    };
+    let offset = info.byte_offset as usize;
+    let end_of_parameter = offset + parameter.len();
+
+    if end_of_parameter > *max_size as usize {
+        // TODO: tracing for which parameter was too big?
+        return Err(Error::ParameterTooLarge);
+    }
+
+    if parameter_area.len() < end_of_parameter {
+        parameter_area.resize(end_of_parameter, 0);
+    }
+
+    parameter_area[offset..end_of_parameter].copy_from_slice(parameter);
+    Ok(())
+}
+
+///
+/// Load the given IGVM file to guest memory.
+/// Right now it only supports SNP based isolation
+///
+pub fn load_igvm(
+    mut file: &std::fs::File,
+    memory_manager: Arc<Mutex<MemoryManager>>,
+    cpu_manager: Arc<Mutex<CpuManager>>,
+    mem_regions: Vec<ArchMemRegion>,
+    proc_count: u32,
+    cmdline: &str,
```

Comment:
is this cmdline argument needed? isn't it part of the igvm file itself?

Reply:
If additional cmdline is provided by the VMM